### PR TITLE
ICU-21502 Adds status checks to test to prevent segmentation fault when

### PR DESCRIPTION
**Pull request by gnrunge**
test runs with ICU stub data only.

##### Checklist

- [ ] Issue filed: https://unicode-org.atlassian.net/browse/ICU-21502
- [ ] Updated PR title and link in previous line to include Issue number
- [ ] Issue accepted
- [ ] Tests included
- [ ] Documentation is changed or added

### CODE REVIEW

File: icu4c/source/test/intltest/units_test.cpp [hunk 1]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
...
+          return;
+        }
         ComplexUnitsConverter converter(inputImpl, outputImpl, conversionRates, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
 
         auto actual = converter.convert(testCase.inputValue, nullptr, status);
+        if (U_FAILURE(status)) {
+          return;
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:725]:
```suggestion
            return;
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 2]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
...
+        }
         MeasureUnitImpl outputImpl = MeasureUnitImpl::forIdentifier(testCase.output, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
         ComplexUnitsConverter converter(inputImpl, outputImpl, conversionRates, status);
+        if (U_FAILURE(status)) {
+          return;
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:720]:
```suggestion
            return;
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 3]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
 
     for (const auto &testCase : testCases) {
         MeasureUnitImpl inputImpl = MeasureUnitImpl::forIdentifier(testCase.input, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
         MeasureUnitImpl outputImpl = MeasureUnitImpl::forIdentifier(testCase.output, status);
+        if (U_FAILURE(status)) {
+          return;
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:716]:
```suggestion
            return;
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 4]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
 
     for (const auto &testCase : testCases) {
         MeasureUnitImpl inputImpl = MeasureUnitImpl::forIdentifier(testCase.input, status);
+        if (U_FAILURE(status)) {
+          return;
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:712]:
```suggestion
            return;
```

**Review by hugovdm — APPROVED** [review]
Thanks for this! I'll consider it a good workaround, and will dig deeper into the guts of the segfaulting code (which should handle status correctly).

Feel free to ignore the space suggestions, so you can simply merge! I can deal with them when I dig in...

**Comment by hugovdm** [comment]
This might impact ICU-21432 too? Or a similar workaround makes sense to get the tests passing. (I'm also happy to repurpose ICU-21432 to be about internals, or to file a new ticket. Apologies that I've not been able ot tackle it yet! Soon!)

File: icu4c/source/test/intltest/units_test.cpp (hunk 2 above)
Comment (hugovdm) [reply icu4c/source/test/intltest/units_test.cpp:720]: Better:

```suggestion
        if (status.errIfFailureAndReset()) {
            continue;
        }
```

File: icu4c/source/test/intltest/units_test.cpp (hunk 3 above)
Comment (hugovdm) [reply icu4c/source/test/intltest/units_test.cpp:716]: Better:

```suggestion
        if (status.errIfFailureAndReset()) {
            continue;
        }
```

File: icu4c/source/test/intltest/units_test.cpp (hunk 4 above)
Comment (hugovdm) [reply icu4c/source/test/intltest/units_test.cpp:712]: Better:

```suggestion
        if (status.errIfFailureAndReset()) {
            continue;
        }
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 5]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
...
+        }
         ComplexUnitsConverter converter(inputImpl, outputImpl, conversionRates, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
 
         auto actual = converter.convert(testCase.inputValue, nullptr, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:726]:
```suggestion
        if (status.errIfFailureAndReset()) {
            continue;
        }
        if (actual.length() < testCase.expectedCount) {
            errln("converter.convert(...) returned too few Measures");
            continue;
        }
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 6]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
 
     for (const auto &testCase : testCases) {
         MeasureUnitImpl inputImpl = MeasureUnitImpl::forIdentifier(testCase.input, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:713]:
```suggestion
        if (status.errIfFailureAndReset()) {
            continue;
        }
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 7]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
...
         MeasureUnitImpl outputImpl = MeasureUnitImpl::forIdentifier(testCase.output, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:717]:
```suggestion
        if (status.errIfFailureAndReset()) {
            continue;
        }
```

File: icu4c/source/test/intltest/units_test.cpp [hunk 8]
```diff
@@ -708,10 +708,22 @@ void UnitsTest::testComplexUnitsConverterSorting() {
...
         ComplexUnitsConverter converter(inputImpl, outputImpl, conversionRates, status);
+        if (U_FAILURE(status)) {
+          return;
+        }
```
Comment (hugovdm) [inline icu4c/source/test/intltest/units_test.cpp:721]:
```suggestion
        if (status.errIfFailureAndReset()) {
            continue;
        }
```

File: icu4c/source/test/intltest/units_test.cpp (hunk 3 above)
Comment (younies) [reply icu4c/source/test/intltest/units_test.cpp:716]: +1

File: icu4c/source/test/intltest/units_test.cpp (hunk 4 above)
Comment (younies) [reply icu4c/source/test/intltest/units_test.cpp:712]: +1

**Review by hugovdm — CHANGES_REQUESTED** [review]
Proposing some adjustments to reveal rather than hide errors (errors that shouldn't happen of course).

File: icu4c/source/test/intltest/units_test.cpp (hunk 3 above)
Comment (hugovdm) [reply icu4c/source/test/intltest/units_test.cpp:716]: (Redid my suggestions as I did them wrong - couldn't be directly applied. Deleted the old ones.)

File: icu4c/source/test/intltest/units_test.cpp (hunk 5 above)
Comment (younies) [reply icu4c/source/test/intltest/units_test.cpp:726]: +1

@gnrunge , you can also add to the message the actual length and the expected length. (optional)

**Review by younies — APPROVED** [review]

**Review by hugovdm — COMMENTED** [review]
A quick PR for my proposed changes, so I can get CI tests to run: https://github.com/gnrunge/icu/pull/1

Up to you how you'd like us to handle the suggestions! (Fastest round-trip is if I squash and provide an upstream PR?)

File: icu4c/source/test/intltest/units_test.cpp (hunk 3 above)
Comment (gnrunge) [reply icu4c/source/test/intltest/units_test.cpp]: Worked in the review comments.

File: icu4c/source/test/intltest/units_test.cpp (hunk 4 above)
Comment (gnrunge) [reply icu4c/source/test/intltest/units_test.cpp]: Done.

File: icu4c/source/test/intltest/units_test.cpp (hunk 5 above)
Comment (gnrunge) [reply icu4c/source/test/intltest/units_test.cpp]: Done.

File: icu4c/source/test/intltest/units_test.cpp (hunk 6 above)
Comment (gnrunge) [reply icu4c/source/test/intltest/units_test.cpp]: Done.

File: icu4c/source/test/intltest/units_test.cpp (hunk 8 above)
Comment (gnrunge) [reply icu4c/source/test/intltest/units_test.cpp]: Done.

File: icu4c/source/test/intltest/units_test.cpp (hunk 7 above)
Comment (gnrunge) [reply icu4c/source/test/intltest/units_test.cpp]: Done.

**Review by hugovdm — APPROVED** [review]

**Pull request merged**